In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

In [3]:
import tensorflow as tf

# Helper libraries
import numpy as np
import matplotlib.pyplot as plt

print(tf.__version__)

2.3.0


In [4]:
fashion_mnist = tf.keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

In [5]:
train_images = train_images / 255.0

test_images = test_images / 255.0

## Model #1:  Original Model -  1. adam, dense128, dense10

In [6]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10)
])

In [7]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

In [8]:
model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 1s 581us/step - loss: 0.4910 - accuracy: 0.8266
Epoch 2/10
1875/1875 [==============================] - 1s 613us/step - loss: 0.3698 - accuracy: 0.8672
Epoch 3/10
1875/1875 [==============================] - 1s 593us/step - loss: 0.3365 - accuracy: 0.8771
Epoch 4/10
1875/1875 [==============================] - 1s 582us/step - loss: 0.3097 - accuracy: 0.8865
Epoch 5/10
1875/1875 [==============================] - 1s 577us/step - loss: 0.2947 - accuracy: 0.8909
Epoch 6/10
1875/1875 [==============================] - 1s 580us/step - loss: 0.2776 - accuracy: 0.8986
Epoch 7/10
1875/1875 [==============================] - 1s 587us/step - loss: 0.2668 - accuracy: 0.8998
Epoch 8/10
1875/1875 [==============================] - 1s 643us/step - loss: 0.2563 - accuracy: 0.9049
Epoch 9/10
1875/1875 [==============================] - 1s 613us/step - loss: 0.2481 - accuracy: 0.9074
Epoch 10/10
1875/1875 [==============================] - 1s 649u

In [9]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

313/313 - 0s - loss: 0.3445 - accuracy: 0.8837

Test accuracy: 0.8837000131607056


## Model #2 -  2. adam, dense10, dropout(.2), dense10

In [10]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(10, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [11]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

In [12]:
model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 1s 577us/step - loss: 1.0347 - accuracy: 0.6003
Epoch 2/10
1875/1875 [==============================] - 1s 542us/step - loss: 0.8170 - accuracy: 0.6881
Epoch 3/10
1875/1875 [==============================] - 1s 539us/step - loss: 0.7835 - accuracy: 0.7058
Epoch 4/10
1875/1875 [==============================] - 1s 537us/step - loss: 0.7608 - accuracy: 0.7155
Epoch 5/10
1875/1875 [==============================] - 1s 538us/step - loss: 0.7440 - accuracy: 0.7230
Epoch 6/10
1875/1875 [==============================] - 1s 726us/step - loss: 0.7333 - accuracy: 0.7273
Epoch 7/10
1875/1875 [==============================] - 1s 549us/step - loss: 0.7284 - accuracy: 0.7325
Epoch 8/10
1875/1875 [==============================] - 1s 553us/step - loss: 0.7195 - accuracy: 0.7340
Epoch 9/10
1875/1875 [==============================] - 1s 547us/step - loss: 0.7139 - accuracy: 0.7375
Epoch 10/10
1875/1875 [==============================] - 1s 540u

In [13]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

313/313 - 0s - loss: 0.5251 - accuracy: 0.8255

Test accuracy: 0.8255000114440918


## Model #3: Model #1 performed better:  1. adagrad

In [14]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10)
])

In [15]:
model.compile(optimizer='adagrad',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

In [16]:
model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 1s 657us/step - loss: 1.0692 - accuracy: 0.6680
Epoch 2/10
1875/1875 [==============================] - 1s 648us/step - loss: 0.7241 - accuracy: 0.7650
Epoch 3/10
1875/1875 [==============================] - 1s 645us/step - loss: 0.6510 - accuracy: 0.7895
Epoch 4/10
1875/1875 [==============================] - 1s 641us/step - loss: 0.6115 - accuracy: 0.8026
Epoch 5/10
1875/1875 [==============================] - 1s 664us/step - loss: 0.5852 - accuracy: 0.8098
Epoch 6/10
1875/1875 [==============================] - 1s 645us/step - loss: 0.5662 - accuracy: 0.8159
Epoch 7/10
1875/1875 [==============================] - 1s 648us/step - loss: 0.5512 - accuracy: 0.8193
Epoch 8/10
1875/1875 [==============================] - 1s 650us/step - loss: 0.5392 - accuracy: 0.8227
Epoch 9/10
1875/1875 [==============================] - 1s 678us/step - loss: 0.5290 - accuracy: 0.8254
Epoch 10/10
1875/1875 [==============================] - 1s 650u

In [17]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

313/313 - 0s - loss: 0.5417 - accuracy: 0.8179

Test accuracy: 0.8179000020027161


## Model #4: Model #1 performed better:  2. sgd

In [18]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10)
])

In [19]:
model.compile(optimizer='sgd',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

In [20]:
model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 1s 622us/step - loss: 0.7440 - accuracy: 0.7596
Epoch 2/10
1875/1875 [==============================] - 1s 610us/step - loss: 0.5165 - accuracy: 0.8235
Epoch 3/10
1875/1875 [==============================] - 1s 611us/step - loss: 0.4737 - accuracy: 0.8371
Epoch 4/10
1875/1875 [==============================] - 1s 611us/step - loss: 0.4489 - accuracy: 0.8442
Epoch 5/10
1875/1875 [==============================] - 1s 605us/step - loss: 0.4321 - accuracy: 0.8515
Epoch 6/10
1875/1875 [==============================] - 1s 607us/step - loss: 0.4183 - accuracy: 0.8554
Epoch 7/10
1875/1875 [==============================] - 1s 613us/step - loss: 0.4062 - accuracy: 0.8592
Epoch 8/10
1875/1875 [==============================] - 1s 615us/step - loss: 0.3960 - accuracy: 0.8618
Epoch 9/10
1875/1875 [==============================] - 1s 617us/step - loss: 0.3873 - accuracy: 0.8651
Epoch 10/10
1875/1875 [==============================] - 1s 616u

In [21]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

313/313 - 0s - loss: 0.4127 - accuracy: 0.8545

Test accuracy: 0.8544999957084656


### Results

<p>Model #1<br>
options: relu, adam<br> 
train time: 1 second<br> 
train acc: 0.9096<br>
test acc 0.8837<br>
time: 1s<br>
dense128, dense10 - same as original<br>

Model #2<br>
options: relu, adam<br> 
train time: 1 second<br> 
train acc: 0.7399<br>
test acc 0.8255<br>
time: 1s<br>
dense10, dropout(.2), dense10<br>

Model #3<br>
options: relu, adagrad<br> 
train time: 1 second<br> 
train acc: 0.8282<br>
test acc 0.8179<br>
time: 1s<br>
dense128, dense10 - same as Model #1<br>

Model #4<br>
options: relu, sdg<br> 
train time: 1 second<br> 
train acc: 0.8677<br>
test acc 0.8545<br>
time: 1s<br>
dense128, dense10 - same as Model #1<br>